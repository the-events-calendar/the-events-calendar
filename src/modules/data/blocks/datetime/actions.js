/**
 * Internal dependencies
 */
import * as types from './types';

export const setNaturalLanguageLabel = ( label, meta = {} ) => ( {
	type: types.SET_NATURAL_LANGUAGE_LABEL,
	payload: {
		label,
	},
	meta,
} );

export const setDateRange = ( payload, meta = {} ) => ( {
	type: types.SET_DATE_RANGE,
	payload,
	meta,
} );

export const setStartDateTime = ( start, meta = {} ) => ( {
	type: types.SET_START_DATE_TIME,
	payload: {
		start,
	},
	meta,
} );

export const setEndDateTime = ( end, meta = {} ) => ( {
	type: types.SET_END_DATE_TIME,
	payload: {
		end,
	},
	meta,
} );

export const setStartTime = ( start, meta = {} ) => ( {
	type: types.SET_START_TIME,
	payload: {
		start,
	},
	meta,
} );

export const setEndTime = ( end, meta = {} ) => ( {
	type: types.SET_END_TIME,
	payload: {
		end,
	},
	meta,
} );

export const setStartTimeInput = ( startTimeInput ) => ( {
	type: types.SET_START_TIME_INPUT,
	payload: {
		startTimeInput,
	},
} );

export const setEndTimeInput = ( endTimeInput ) => ( {
	type: types.SET_END_TIME_INPUT,
	payload: {
		endTimeInput,
	},
} );

export const setSeparatorDate = ( separator ) => ( {
	type: types.SET_SEPARATOR_DATE,
	payload: {
		separator,
	},
} );

export const setSeparatorTime = ( separator ) => ( {
	type: types.SET_SEPARATOR_TIME,
	payload: {
		separator,
	},
} );

export const setAllDay = ( allDay ) => ( {
	type: types.SET_ALL_DAY,
	payload: {
		allDay,
	},
} );

export const setMultiDay = ( multiDay, meta = {} ) => ( {
	type: types.SET_MULTI_DAY,
	payload: {
		multiDay,
	},
	meta,
} );

export const setTimeZone = ( timeZone ) => ( {
	type: types.SET_TIME_ZONE,
	payload: {
		timeZone,
	},
} );

export const setTimeZoneVisibility = ( show ) => ( {
	type: types.SET_TIMEZONE_VISIBILITY,
	payload: {
		show,
	},
} );

export const allowEdits = () => ( {
	type: types.SET_DATETIME_BLOCK_EDITABLE_STATE,
	payload: {
		isEditable: true,
	},
} );

export const disableEdits = () => ( {
	type: types.SET_DATETIME_BLOCK_EDITABLE_STATE,
	payload: {
		isEditable: false,
	},
} );
