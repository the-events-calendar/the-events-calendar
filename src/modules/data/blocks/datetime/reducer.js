/**
 * External dependencies
 */
import { __ } from '@wordpress/i18n';
import moment from 'moment/moment';

/**
 * Internal dependencies
 */
import {
	globals,
	date,
	moment as momentUtil,
	time,
} from '@moderntribe/common/utils';
import * as types from './types';

const { parseFormats, toTime } = momentUtil;

export const defaultStateToMetaMap = {
	start: '_EventStartDate',
	end: '_EventEndDate',
	dateTimeSeparator: '_EventDateTimeSeparator',
	timeRangeSeparator: '_EventTimeRangeSeparator',
	allDay: '_EventAllDay',
	timeZone: '_EventTimezone',
};

export const setInitialState = ( data ) => {
	const { meta } = data;

	Object.keys( defaultStateToMetaMap ).forEach( ( key ) => {
		const metaKey = defaultStateToMetaMap[ key ];
		if ( meta.hasOwnProperty( metaKey ) ) {
			DEFAULT_STATE[ key ] = meta[ metaKey ];
		}
	} );

	const { start, end } = DEFAULT_STATE;

	DEFAULT_STATE.startTimeInput = toTime( parseFormats( start ) );
	DEFAULT_STATE.endTimeInput = toTime( parseFormats( end ) );
	DEFAULT_STATE.naturalLanguageLabel = date.rangeToNaturalLanguage( start, end );
	DEFAULT_STATE.multiDay = ! momentUtil.isSameDay( momentUtil.toMoment( start ), momentUtil.toMoment( end ) );
};

const defaultStartTime = globals.defaultTimes().start ? globals.defaultTimes().start : '08:00:00';
const defaultEndTime = globals.defaultTimes().end ? globals.defaultTimes().end : '17:00:00';
const defaultStartTimeSeconds = time.toSeconds( defaultStartTime, time.TIME_FORMAT_HH_MM_SS );
const defaultEndTimeSeconds = time.toSeconds( defaultEndTime, time.TIME_FORMAT_HH_MM_SS );

export const defaultStartMoment = moment().startOf( 'day' ).seconds( defaultStartTimeSeconds );
export const defaultEndMoment = moment().startOf( 'day' ).seconds( defaultEndTimeSeconds );

const defaultStartDateTime = momentUtil.toDateTime( defaultStartMoment );
const defaultEndDateTime = momentUtil.toDateTime( defaultEndMoment );

export const DEFAULT_STATE = {
	start: defaultStartDateTime,
	end: defaultEndDateTime,
	startTimeInput: momentUtil.toTime( defaultStartMoment ),
	endTimeInput: momentUtil.toTime( defaultEndMoment ),
	naturalLanguageLabel: date.rangeToNaturalLanguage( defaultStartDateTime, defaultEndDateTime ),
	dateTimeSeparator: globals.settings().dateTimeSeparator
		? globals.settings().dateTimeSeparator
		: __( '@', 'the-events-calendar' ),
	timeRangeSeparator: globals.settings().timeRangeSeparator
		? globals.settings().timeRangeSeparator
		: __( '-', 'the-events-calendar' ),
	allDay: false,
	multiDay: false,
	timeZone: globals.timezone().timeZone ? globals.timezone().timeZone : date.FORMATS.TIMEZONE.string,
	showTimeZone: false,
	isEditable: true,
};

export default ( state = DEFAULT_STATE, action ) => {
	switch ( action.type ) {
		case types.SET_DATETIME_BLOCK_EDITABLE_STATE:
			return {
				...state,
				isEditable: action.payload.isEditable,
			};
		case types.SET_START_DATE_TIME:
			return {
				...state,
				start: action.payload.start,
			};
		case types.SET_END_DATE_TIME:
			return {
				...state,
				end: action.payload.end,
			};
		case types.SET_START_TIME_INPUT:
			return {
				...state,
				startTimeInput: action.payload.startTimeInput,
			};
		case types.SET_END_TIME_INPUT:
			return {
				...state,
				endTimeInput: action.payload.endTimeInput,
			};
		case types.SET_NATURAL_LANGUAGE_LABEL:
			return {
				...state,
				naturalLanguageLabel: action.payload.label,
			};
		case types.SET_ALL_DAY:
			return {
				...state,
				allDay: action.payload.allDay,
			};
		case types.SET_MULTI_DAY:
			return {
				...state,
				multiDay: action.payload.multiDay,
			};
		case types.SET_SEPARATOR_DATE:
			return {
				...state,
				dateTimeSeparator: action.payload.separator,
			};
		case types.SET_SEPARATOR_TIME:
			return {
				...state,
				timeRangeSeparator: action.payload.separator,
			};
		case types.SET_TIME_ZONE:
			return {
				...state,
				timeZone: action.payload.timeZone,
			};
		case types.SET_TIMEZONE_VISIBILITY:
			return {
				...state,
				showTimeZone: action.payload.show,
			};
		default:
			return state;
	}
};
