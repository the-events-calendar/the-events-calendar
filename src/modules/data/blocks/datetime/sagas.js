/**
 * External Dependencies
 */
import { put, take, select, call, all } from 'redux-saga/effects';
import { applyFilters } from '@wordpress/hooks';

/**
 * Internal dependencies
 */
import {
	types,
	selectors,
	actions,
} from '@moderntribe/events/data/blocks/datetime';
import {
	date as dateUtil,
	moment as momentUtil,
	time as timeUtil,
} from '@moderntribe/common/utils';

const {
	HALF_HOUR_IN_SECONDS,
	DAY_IN_SECONDS,
	HOUR_IN_SECONDS,
	MINUTE_IN_SECONDS,
} = timeUtil;

//
// ──────────────────────────────────────────────────── I ──────────
//   :::::: D E R I V E : :  :   :    :     :        :          :
// ──────────────────────────────────────────────────────────────
//

/**
 * Convert current start and end date into moments
 *
 * @export
 * @since 4.7
 * @returns {Object} {start, end}
 */
export function* deriveMomentsFromDates() {
	const dates = yield all( {
		start: select( selectors.getStart ),
		end: select( selectors.getEnd ),
	} );

	return yield all( {
		start: call( momentUtil.toMoment, dates.start ),
		end: call( momentUtil.toMoment, dates.end ),
	} );
}

/**
 * Convert current start and end date into seconds
 *
 * @export
 * @since 4.7
 * @returns {Object} {start, end}
 */
export function* deriveSecondsFromDates() {
	const moments = yield call( deriveMomentsFromDates );

	const time = yield all( {
		start: call( momentUtil.toDatabaseTime, moments.start ),
		end: call( momentUtil.toDatabaseTime, moments.end ),
	} );

	return yield all( {
		start: call( timeUtil.toSeconds, time.start ),
		end: call( timeUtil.toSeconds, time.end ),
	} );
}

//
// ──────────────────────────────────────────────────────────────────── II ──────────
//   :::::: H U M A N   R E A D A B L E : :  :   :    :     :        :          :
// ──────────────────────────────────────────────────────────────────────────────
//

/**
 * Set the human readable label into the store, based on a start and end date to generate a new label based on those
 *
 * @since 0.3.1-alpha
 *
 * @param {object} dates An object that represents the start / end date
 * @returns {IterableIterator<*>}
 */
export function* setHumanReadableLabel( dates = {} ) {
	const currentLabel = yield select( selectors.getNaturalLanguageLabel );

	if ( currentLabel === '' ) {
		return;
	}

	const updatedLabel = yield call( dateUtil.rangeToNaturalLanguage, dates.start, dates.end );

	if ( currentLabel !== updatedLabel ) {
		yield put( actions.setNaturalLanguageLabel( updatedLabel ) );
	}
}

/**
 * Set the humman readable label from an action that sets the date on the event either the start or end date
 * first selects as default value current start and end dates after that based on the action type selects the
 * date to be replaced or the new date to be set before doing the conversion into text.
 *
 * @since 0.3.1-alpha
 *
 * @param {object} action Dispateched by the component and watched by this generator
 * @returns {IterableIterator<*>}
 */
export function* setHumanReadableFromDate( action ) {
	const dates = {
		start: yield select( selectors.getStart ),
		end: yield select( selectors.getEnd ),
	};

	if ( action.type === types.SET_END_DATE_TIME ) {
		dates.end = action.payload.end;
	} else {
		dates.start = action.payload.start;
	}
	yield call( setHumanReadableLabel, dates );
}

/**
 * Generator used to reset the label of the input using the current values on the start and end date, useful when
 * the user enters a data that is not valid or can't be parsed into a valid date so the input always reflects a valid
 * date based on the current values set into the dates of the event.
 *
 * @since 0.3.1-alpha
 *
 * @returns {IterableIterator<*>}
 */
export function* resetNaturalLanguageLabel() {
	const dates = yield all( {
		start: select( selectors.getStart ),
		end: select( selectors.getEnd ),
	} );
	yield call( setHumanReadableLabel, dates );
}

/**
 * Fired when the Human Readable has a change() event fired on the input
 * when the input is handling a new change() events
 *
 * @since 0.3.1-alpha
 *
 * @returns {IterableIterator<*>}
 */
export function* onHumanReadableChange( action ) {
	const label = yield select( selectors.getNaturalLanguageLabel );
	const { start, end } = yield call( dateUtil.labelToDate, label );

	if ( start === null && end === null ) {
		yield call( resetNaturalLanguageLabel );
	} else {
		const moments = yield all( {
			start: call( momentUtil.toMoment, start ),
			end: call( momentUtil.toMoment, end || start ),
		} );

		const result = yield call( momentUtil.adjustStart, moments.start, moments.end );

		const isMultiDay = ! ( yield call( momentUtil.isSameDay, result.start, result.end ) );

		/**
		 * @todo: move this to common utils in moment.js
		 */
		const isAllDay = ! isMultiDay && '00:00' === moments.start.format( 'HH:mm' ) && '23:59' === moments.end.format( 'HH:mm' );

		const dates = yield all( {
			start: call( momentUtil.toDateTime, result.start ),
			end: call( momentUtil.toDateTime, result.end ),
		} );

		yield all( [
			call( action.meta.setAttributes, { start: dates.start } ),
			call( action.meta.setAttributes, { end: dates.end } ),
			call( action.meta.setAttributes, { allDay: isAllDay } ),
		] );

		yield all( [
			put( actions.setStartDateTime( dates.start ) ),
			put( actions.setEndDateTime( dates.end ) ),
			put( actions.setMultiDay( isMultiDay ) ),
			put( actions.setAllDay( isAllDay ) ),
		] );
	}
}

//
// ────────────────────────────────────────────────────────────────────── III ──────────
//   :::::: C H A N G E   H A N D L E R S : :  :   :    :     :        :          :
// ────────────────────────────────────────────────────────────────────────────────
//

/**
 * Handle date range changes on calendar
 *
 * @export
 * @since 4.7
 * @param {Object} action Payload with to and from
 */
export function* handleDateRangeChange( action ) {
	const { to, from } = action.payload;
	const moments = yield call( deriveMomentsFromDates );

	const rangeMoments = yield all( {
		from: call( momentUtil.toMoment, from ),
		to: call( momentUtil.toMoment, to || from ),
	} );

	// NOTE: Mutation
	yield all( {
		start: call( momentUtil.replaceDate, moments.start, rangeMoments.from ),
		end: call( momentUtil.replaceDate, moments.end, rangeMoments.to ),
	} );

	const result = yield call( momentUtil.adjustStart, moments.start, moments.end );

	const dates = yield all( {
		start: call( momentUtil.toDateTime, result.start ),
		end: call( momentUtil.toDateTime, result.end ),
	} );

	yield all( [
		call( action.meta.setAttributes, { start: dates.start } ),
		call( action.meta.setAttributes, { end: dates.end } ),
	] );

	yield all( [
		put( actions.setStartDateTime( dates.start ) ),
		put( actions.setEndDateTime( dates.end ) ),
	] );
}

/**
 * Prevents end time from being before start time.
 * Should only prevent when not a multi-day event.
 *
 * @export
 * @since 4.7
 * @param {Object} action Payload with seconds in start or end key (when time change)
 */
export function* preventEndTimeBeforeStartTime( action ) {
	const isMultiDay = yield select( selectors.getMultiDay );
	// Do not proceed when multi-day
	if ( isMultiDay ) {
		return;
	}

	const seconds = yield call( deriveSecondsFromDates );
	// Prevent only date changes from using payload
	if ( [ types.SET_END_TIME, types.SET_START_TIME ].includes( action.type ) ) {
		// Update seconds to use payload
		// NOTE: Mutation
		yield call( [ Object, 'assign' ], seconds, action.payload );
	}

	// If end time is earlier than start time, fix time
	if ( seconds.end < seconds.start ) {
		// If there is less than half an hour left in the day, roll back one hour
		if ( seconds.start + HALF_HOUR_IN_SECONDS >= DAY_IN_SECONDS ) {
			seconds.start -= HOUR_IN_SECONDS;
		}

		seconds.end = seconds.start + HALF_HOUR_IN_SECONDS;

		const moments = yield call( deriveMomentsFromDates );

		// NOTE: Mutation
		yield all( {
			start: call( momentUtil.setTimeInSeconds, moments.start, seconds.start ),
			end: call( momentUtil.setTimeInSeconds, moments.end, seconds.end ),
		} );

		const dates = yield all( {
			start: call( momentUtil.toDateTime, moments.start ),
			end: call( momentUtil.toDateTime, moments.end ),
		} );

		yield all( [
			call( action.meta.setAttributes, { start: dates.start } ),
			call( action.meta.setAttributes, { end: dates.end } ),
		] );

		yield all( [
			put( actions.setStartDateTime( dates.start ) ),
			put( actions.setEndDateTime( dates.end ) ),
		] );
	}
}

/**
 * Prevents start time from appearing ahead of end time.
 * Should only prevent when not a multi-day event.
 *
 * @export
 * @since 4.7
 * @param {Object} action Payload with seconds in start or end key (when time change)
 */
export function* preventStartTimeAfterEndTime( action ) {
	const isMultiDay = yield select( selectors.getMultiDay );
	// Do not proceed when multi-day
	if ( isMultiDay ) {
		return;
	}

	const seconds = yield call( deriveSecondsFromDates );

	// Prevent only date changes from using payload
	if ( [ types.SET_END_TIME, types.SET_START_TIME ].includes( action.type ) ) {
		// Update seconds to use payload
		// NOTE: Mutation
		yield call( [ Object, 'assign' ], seconds, action.payload );
	}

	if ( seconds.start > seconds.end ) {
		seconds.start = Math.max( seconds.end - HALF_HOUR_IN_SECONDS, 0 );
		seconds.end = Math.max( seconds.start + MINUTE_IN_SECONDS, seconds.end );

		const moments = yield call( deriveMomentsFromDates );

		// NOTE: Mutation
		yield all( {
			start: call( momentUtil.setTimeInSeconds, moments.start, seconds.start ),
			end: call( momentUtil.setTimeInSeconds, moments.end, seconds.end ),
		} );

		const dates = yield all( {
			start: call( momentUtil.toDateTime, moments.start ),
			end: call( momentUtil.toDateTime, moments.end ),
		} );

		yield all( [
			call( action.meta.setAttributes, { start: dates.start } ),
			call( action.meta.setAttributes, { end: dates.end } ),
		] );

		yield all( [
			put( actions.setStartDateTime( dates.start ) ),
			put( actions.setEndDateTime( dates.end ) ),
		] );
	}
}
/**
 * Handles all-day payloads. Set start and end time to be `00:00` and `23:59`
 *
 * @export
 * @since 4.7
 */
export function* setAllDay( action ) {
	const moments = yield call( deriveMomentsFromDates );

	// NOTE: Mutation
	yield all( {
		start: call( momentUtil.setTimeInSeconds, moments.start, 0 ),
		end: call( momentUtil.setTimeInSeconds, moments.end, timeUtil.DAY_IN_SECONDS - 1 ),
	} );

	const dates = yield all( {
		start: call( momentUtil.toDateTime, moments.start ),
		end: call( momentUtil.toDateTime, moments.end ),
	} );

	yield all( [
		call( action.meta.setAttributes, { start: dates.start } ),
		call( action.meta.setAttributes, { end: dates.end } ),
		call( action.meta.setAttributes, { allDay: true } ),
	] );

	yield all( [
		put( actions.setStartDateTime( dates.start ) ),
		put( actions.setEndDateTime( dates.end ) ),
		put( actions.setAllDay( true ) ),
	] );
}

/**
 * Handles multi-day toggling
 *
 * @export
 * @since 4.7
 * @param {Object} action Payload with multiDay
 */
export function* handleMultiDay( action ) {
	const isMultiDay = action.payload.multiDay;
	const { start, end } = yield call( deriveMomentsFromDates );

	if ( isMultiDay ) {
		const RANGE_DAYS = yield call( applyFilters, 'tec.datetime.defaultRange', 3 );
		// NOTE: Mutation
		yield call( [ end, 'add' ], RANGE_DAYS, 'days' );
		const endDate = yield call( momentUtil.toDateTime, end );

		yield call( action.meta.setAttributes, { end: endDate } );
		yield put( actions.setEndDateTime( endDate ) );
	} else {
		const newEnd = yield call( momentUtil.replaceDate, end, start );
		const result = yield call( momentUtil.adjustStart, start, newEnd );

		const dates = yield all( {
			start: call( momentUtil.toDateTime, result.start ),
			end: call( momentUtil.toDateTime, result.end ),
		} );

		yield all( [
			call( action.meta.setAttributes, { start: dates.start } ),
			call( action.meta.setAttributes, { end: dates.end } ),
		] );

		yield all( [
			put( actions.setStartDateTime( dates.start ) ),
			put( actions.setEndDateTime( dates.end ) ),
		] );
	}
}

/**
 * Handles event start time changes
 *
 * @export
 * @since 4.7
 * @param {Object} action Payload with start of `all-day` or seconds
 */
export function* handleStartTimeChange( action ) {
	if ( action.payload.start === 'all-day' ) {
		yield call( setAllDay, action );
	} else {
		// Set All day to false in case they're editing.
		yield call( action.meta.setAttributes, { allDay: false } );
		yield put( actions.setAllDay( false ) );

		const { start } = yield call( deriveMomentsFromDates );
		// NOTE: Mutation
		yield call( momentUtil.setTimeInSeconds, start, action.payload.start );
		const startDate = yield call( momentUtil.toDateTime, start );

		yield call( action.meta.setAttributes, { start: startDate } );
		yield put( actions.setStartDateTime( startDate ) );
	}
}

/**
 * Handles event end time changes
 *
 * @export
 * @since 4.7
 * @param {Object} action Payload with end of `all-day` or seconds
 */
export function* handleEndTimeChange( action ) {
	if ( action.payload.end === 'all-day' ) {
		yield call( setAllDay, action );
	} else {

		// Set All day to false in case they're editing.
		yield call( action.meta.setAttributes, { allDay: false } );
		yield put( actions.setAllDay( false ) );

		const { end } = yield call( deriveMomentsFromDates );
		// NOTE: Mutation
		yield call( momentUtil.setTimeInSeconds, end, action.payload.end );
		const endDate = yield call( momentUtil.toDateTime, end );

		yield call( action.meta.setAttributes, { end: endDate } );
		yield put( actions.setEndDateTime( endDate ) );
	}
}

/**
 * Sets start time input
 *
 * @export
 * @since 4.7.2
 */
export function* setStartTimeInput() {
	const { start } = yield call( deriveMomentsFromDates );
	const startInput = yield call( momentUtil.toTime, start );
	yield put( actions.setStartTimeInput( startInput ) );
}

/**
 * Sets end time input
 *
 * @export
 * @since 4.7.2
 */
export function* setEndTimeInput() {
	const { end } = yield call( deriveMomentsFromDates );
	const endInput = yield call( momentUtil.toTime, end );
	yield put( actions.setEndTimeInput( endInput ) );
}

/**
 * Handle flow changes based on action type
 *
 * @export
 * @since 4.7
 * @param {Object} action Action taken
 */
export function* handler( action ) {
	switch ( action.type ) {
		case types.SET_DATE_RANGE:
			yield call( handleDateRangeChange, action );
			yield call( resetNaturalLanguageLabel );
			break;

		case types.SET_START_DATE_TIME:
			yield call( preventEndTimeBeforeStartTime, action );
			yield call( setHumanReadableFromDate, action );
			break;

		case types.SET_END_DATE_TIME:
			yield call( preventStartTimeAfterEndTime, action );
			yield call( setHumanReadableFromDate, action );
			break;

		case types.SET_START_TIME:
			yield call( handleStartTimeChange, action );
			yield call( preventEndTimeBeforeStartTime, action );
			yield call( setStartTimeInput );
			yield call( setEndTimeInput );
			yield call( resetNaturalLanguageLabel );
			break;

		case types.SET_END_TIME:
			yield call( handleEndTimeChange, action );
			yield call( preventStartTimeAfterEndTime, action );
			yield call( setEndTimeInput );
			yield call( setStartTimeInput );
			yield call( resetNaturalLanguageLabel );
			break;

		case types.SET_MULTI_DAY:
			yield call( handleMultiDay, action );
			yield call( setStartTimeInput );
			yield call( setEndTimeInput );
			yield call( resetNaturalLanguageLabel );
			break;

		case types.SET_NATURAL_LANGUAGE_LABEL:
			yield call( onHumanReadableChange, action );
			yield call( setStartTimeInput );
			yield call( setEndTimeInput );
			break;

		default:
			break;
	}
}

//
// ──────────────────────────────────────────────────────── IV ──────────
//   :::::: W A T C H E R S : :  :   :    :     :        :          :
// ──────────────────────────────────────────────────────────────────
//

/**
 * Watchers of actions and act accordingly to each.
 *
 * @since 0.3.1-alpha
 *
 * @returns {IterableIterator<*|ForkEffect>}
 */
export default function* watchers() {
	// prevent changes from looping infinitely
	while ( true ) {
		const action = yield take( [
			types.SET_DATE_RANGE,
			types.SET_START_DATE_TIME,
			types.SET_END_DATE_TIME,
			types.SET_START_TIME,
			types.SET_END_TIME,
			types.SET_MULTI_DAY,
			types.SET_TIME_ZONE,
			types.SET_NATURAL_LANGUAGE_LABEL,
		] );
		yield call( handler, action );
	}
}
