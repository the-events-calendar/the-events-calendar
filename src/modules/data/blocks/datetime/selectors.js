/**
 * External dependencies
 */
import { createSelector } from 'reselect';

export const datetimeSelector = ( state ) => state.events.blocks.datetime;

export const getStart = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.start,
);

export const getEnd = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.end,
);

export const getStartTimeInput = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.startTimeInput,
);

export const getEndTimeInput = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.endTimeInput,
);

export const getAllDay = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.allDay,
);

export const getMultiDay = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.multiDay,
);

export const getDateSeparator = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.dateTimeSeparator,
);

export const getTimeSeparator = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.timeRangeSeparator,
);

export const getTimeZone = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.timeZone,
);

export const getTimeZoneVisibility = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.showTimeZone,
);

export const getNaturalLanguageLabel = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.naturalLanguageLabel,
);

export const isEditable = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.isEditable,
);

export const getSameStartEnd = createSelector(
	[ datetimeSelector ],
	( datetime ) => datetime.start === datetime.end,
);
