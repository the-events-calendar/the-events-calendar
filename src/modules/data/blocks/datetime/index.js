/**
 * Internal dependencies
 */
import * as reducer from './reducer';
import * as types from './types';
import * as actions from './actions';
import * as selectors from './selectors';
import sagas from './sagas';

export default reducer.default;
export {
	reducer,
	types,
	actions,
	selectors,
	sagas,
};
