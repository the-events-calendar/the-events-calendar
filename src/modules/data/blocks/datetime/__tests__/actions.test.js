/**
 * Internal dependencies
 */
import { actions } from '@moderntribe/events/data/blocks/datetime';

describe( '[STORE] - Datetime actions', () => {
	test( 'Action to set the start time', () => {
		expect( actions.setStartDateTime( 'June 5, 2018 5:00 pm' ) ).toMatchSnapshot();
	} );

	test( 'Set Natural Language Label', () => {
		expect( actions.setNaturalLanguageLabel( '2 weeks from now' ) ).toMatchSnapshot();
	} );

	test( 'Action to set the start time', () => {
		expect( actions.setEndDateTime( 'June 25, 2018 4:00 pm' ) ).toMatchSnapshot();
	} );

	test( 'Action to set the separator date', () => {
		expect( actions.setSeparatorDate( ' -  ' ) ).toMatchSnapshot();
	} );

	test( 'Action to set the separator time', () => {
		expect( actions.setSeparatorTime( ' @ ' ) ).toMatchSnapshot();
	} );

	test( 'Action to set the time zone', () => {
		expect( actions.setTimeZone( 'UTC' ) ).toMatchSnapshot();
	} );

	test( 'Action to set the visibility of the time zone', () => {
		expect( actions.setTimeZoneVisibility( true ) ).toMatchSnapshot();
		expect( actions.setTimeZoneVisibility( false ) ).toMatchSnapshot();
	} );

	test( 'Action to set all day', () => {
		expect( actions.setAllDay( true ) ).toMatchSnapshot();
		expect( actions.setAllDay( false ) ).toMatchSnapshot();
	} );

	test( 'Action to set the multi day', () => {
		expect( actions.setMultiDay( true ) ).toMatchSnapshot();
		expect( actions.setMultiDay( false ) ).toMatchSnapshot();
	} );

	test( 'Set editable', () => {
		expect( actions.allowEdits() ).toMatchSnapshot();
		expect( actions.disableEdits() ).toMatchSnapshot();
	} );

	test( 'setStartTime', () => {
		expect( actions.setStartTime( 6000 ) ).toMatchSnapshot();
	} );
	test( 'setEndTime', () => {
		expect( actions.setEndTime( 6000 ) ).toMatchSnapshot();
	} );
	test( 'setDateRange', () => {
		expect( actions.setDateRange( { from: 0, to: 1 } ) ).toMatchSnapshot();
	} );
	test( 'setStartTimeInput', () => {
		expect( actions.setStartTimeInput( '18:00' ) ).toMatchSnapshot();
	} );
	test( 'setEndTimeInput', () => {
		expect( actions.setEndTimeInput( '18:00' ) ).toMatchSnapshot();
	} );
} );
