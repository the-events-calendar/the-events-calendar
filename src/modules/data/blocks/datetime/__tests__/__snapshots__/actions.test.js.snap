// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[STORE] - Datetime actions Action to set all day 1`] = `
Object {
  "payload": Object {
    "allDay": true,
  },
  "type": "@@MT/EVENTS/SET_ALL_DAY",
}
`;

exports[`[STORE] - Datetime actions Action to set all day 2`] = `
Object {
  "payload": Object {
    "allDay": false,
  },
  "type": "@@MT/EVENTS/SET_ALL_DAY",
}
`;

exports[`[STORE] - Datetime actions Action to set the multi day 1`] = `
Object {
  "meta": Object {},
  "payload": Object {
    "multiDay": true,
  },
  "type": "@@MT/EVENTS/SET_MULTI_DAY",
}
`;

exports[`[STORE] - Datetime actions Action to set the multi day 2`] = `
Object {
  "meta": Object {},
  "payload": Object {
    "multiDay": false,
  },
  "type": "@@MT/EVENTS/SET_MULTI_DAY",
}
`;

exports[`[STORE] - Datetime actions Action to set the separator date 1`] = `
Object {
  "payload": Object {
    "separator": " -  ",
  },
  "type": "@@MT/EVENTS/SET_SEPARATOR_DATE",
}
`;

exports[`[STORE] - Datetime actions Action to set the separator time 1`] = `
Object {
  "payload": Object {
    "separator": " @ ",
  },
  "type": "@@MT/EVENTS/SET_SEPARATOR_TIME",
}
`;

exports[`[STORE] - Datetime actions Action to set the start time 1`] = `
Object {
  "meta": Object {},
  "payload": Object {
    "start": "June 5, 2018 5:00 pm",
  },
  "type": "@@MT/EVENTS/SET_START_DATE_TIME",
}
`;

exports[`[STORE] - Datetime actions Action to set the start time 2`] = `
Object {
  "meta": Object {},
  "payload": Object {
    "end": "June 25, 2018 4:00 pm",
  },
  "type": "@@MT/EVENTS/SET_END_DATE_TIME",
}
`;

exports[`[STORE] - Datetime actions Action to set the time zone 1`] = `
Object {
  "payload": Object {
    "timeZone": "UTC",
  },
  "type": "@@MT/EVENTS/SET_TIME_ZONE",
}
`;

exports[`[STORE] - Datetime actions Action to set the visibility of the time zone 1`] = `
Object {
  "payload": Object {
    "show": true,
  },
  "type": "@@MT/EVENTS/SET_TIMEZONE_VISIBILITY",
}
`;

exports[`[STORE] - Datetime actions Action to set the visibility of the time zone 2`] = `
Object {
  "payload": Object {
    "show": false,
  },
  "type": "@@MT/EVENTS/SET_TIMEZONE_VISIBILITY",
}
`;

exports[`[STORE] - Datetime actions Set Natural Language Label 1`] = `
Object {
  "meta": Object {},
  "payload": Object {
    "label": "2 weeks from now",
  },
  "type": "@@MT/EVENTS/SET_NATURAL_LANGUAGE_LABEL",
}
`;

exports[`[STORE] - Datetime actions Set editable 1`] = `
Object {
  "payload": Object {
    "isEditable": true,
  },
  "type": "@@MT/EVENTS/SET_DATETIME_BLOCK_EDITABLE_STATE",
}
`;

exports[`[STORE] - Datetime actions Set editable 2`] = `
Object {
  "payload": Object {
    "isEditable": false,
  },
  "type": "@@MT/EVENTS/SET_DATETIME_BLOCK_EDITABLE_STATE",
}
`;

exports[`[STORE] - Datetime actions setDateRange 1`] = `
Object {
  "meta": Object {},
  "payload": Object {
    "from": 0,
    "to": 1,
  },
  "type": "@@MT/EVENTS/SET_DATE_RANGE",
}
`;

exports[`[STORE] - Datetime actions setEndTime 1`] = `
Object {
  "meta": Object {},
  "payload": Object {
    "end": 6000,
  },
  "type": "@@MT/EVENTS/SET_END_TIME",
}
`;

exports[`[STORE] - Datetime actions setEndTimeInput 1`] = `
Object {
  "payload": Object {
    "endTimeInput": "18:00",
  },
  "type": "@@MT/EVENTS/SET_END_TIME_INPUT",
}
`;

exports[`[STORE] - Datetime actions setStartTime 1`] = `
Object {
  "meta": Object {},
  "payload": Object {
    "start": 6000,
  },
  "type": "@@MT/EVENTS/SET_START_TIME",
}
`;

exports[`[STORE] - Datetime actions setStartTimeInput 1`] = `
Object {
  "payload": Object {
    "startTimeInput": "18:00",
  },
  "type": "@@MT/EVENTS/SET_START_TIME_INPUT",
}
`;
